### PR TITLE
Uninstall process safely

### LINKED ISSUE #25: Uninstall process safety
Corresponding to `ProcessArtifact`:

1. Check if the process still exists (by it's PID) after signaling policy-changed event, then try to close it.
2. Accept number of tries & sleep amount - try to do the action, then wait and repeat until the action is preformed or you have reached the number of tries.

### PR DESCRIPTION
**Pull request by EladShriki**

### CODE REVIEW

File: Stresser/FileArtifact.h [hunk 1]
```diff
@@ -9,6 +9,6 @@ class FileArtifact : public IArtifact
 	explicit FileArtifact(const std::string& name, const std::string& data);
 	virtual ~FileArtifact();
 
-	virtual void install() const;
-	virtual void uninstall() const;
+	virtual void install();
+	virtual void uninstall();
 };
```
Comment (AvivShabtay) [inline Stresser/FileArtifact.h:14]: Use override keyword instead:
`void install() override;`
`void uninstall() override;`

It tells the compiler that you have override this functions.

File: Stresser/IArtifact.h [hunk 2]
```diff
@@ -10,8 +10,8 @@ class IArtifact
 	std::string getData() const;
 	std::string getName() const;
 
-	virtual void install() const = 0;
-	virtual void uninstall() const = 0;
+	virtual void install() = 0;
+	virtual void uninstall() = 0;
 
 protected:
 	std::string data;
```
Comment (AvivShabtay) [inline Stresser/IArtifact.h]: Add `m_` specifier before class member.

File: Stresser/ProcessArtifact.cpp [hunk 3]
```diff
@@ -3,9 +3,10 @@
 ProcessArtifact::ProcessArtifact(const std::string& name, const std::string& data) : IArtifact(name, data)
 {
     std::wstring tempPath = LocalPcUtils::getTempPath();
-    std::wstring wideFileName = std::wstring(ATL::CA2W(name.c_str()));
+    std::wstring wideFileName = std::wstring(ATL::CA2W(data.c_str()));
```
Comment (AvivShabtay) [inline Stresser/ProcessArtifact.cpp]: Suggestions: Add util in `StringUtils`, thus you can convert from `stringToWString` and from `wstringToString`.

We could use it globally anywhere in the application.

File: Stresser/ProcessArtifact.cpp [hunk 4]
```diff
@@ -3,9 +3,10 @@
...
 
     this->filePath = std::wstring(tempPath + wideFileName);
+    this->pid = 0;
```
Comment (AvivShabtay) [inline Stresser/ProcessArtifact.cpp]: I thing `0` is also used to represent `System` process in Windows.
Suggestion: Maybe better to use negative sign, like: `-1`.

File: Stresser/ProcessArtifact.cpp [hunk 5]
```diff
@@ -38,13 +39,19 @@ void ProcessArtifact::install() const
         throw Win32ErrorCodeException("Failed to create a process!");
     }
 
-    AutoHandle autoProcessHandle(processInfo.hProcess);
+    AutoHandle autoHProcess(processInfo.hProcess);
```
Comment (AvivShabtay) [inline Stresser/ProcessArtifact.cpp]: It is enough name it: `process`  :-)

File: Stresser/ProcessArtifact.cpp [hunk 6]
```diff
@@ -38,13 +39,19 @@ void ProcessArtifact::install() const
...
+    AutoHandle autoHProcess(processInfo.hProcess); 
     AutoHandle autoThreadHandle(processInfo.hThread);
+    this->pid = processInfo.dwProcessId;
 }
 
-void ProcessArtifact::uninstall() const
+void ProcessArtifact::uninstall()
 {
-    Sleep(2000);
+    const AutoHandle autoProcess(OpenProcess(SYNCHRONIZE, FALSE, this->pid));
```
Comment (AvivShabtay) [inline Stresser/ProcessArtifact.cpp]: It is enough name it: process :-)
(Clean code and all the this shit...)

Comment (AvivShabtay) [inline Stresser/ProcessArtifact.cpp]: Please document `SYNCHRONIZE` - I think you used it to solve the problem with data races (???)

File: Stresser/ProcessArtifact.h [hunk 7]
```diff
@@ -6,17 +6,19 @@
...
 	virtual ~ProcessArtifact();
 
-	virtual void install() const;
-	virtual void uninstall() const;
+	virtual void install();
+	virtual void uninstall();
 
 private:
 	std::wstring filePath;
+	DWORD pid;
```
Comment (AvivShabtay) [inline Stresser/ProcessArtifact.h]: Add m_ specifier before class member.

File: Stresser/ProcessArtifact.h [hunk 8]
```diff
@@ -6,17 +6,19 @@
...
 class ProcessArtifact : public IArtifact
 {
 public:
 	explicit ProcessArtifact(const std::string& name, const std::string& data);
 	virtual ~ProcessArtifact();
 
-	virtual void install() const;
-	virtual void uninstall() const;
+	virtual void install();
+	virtual void uninstall();
```
Comment (AvivShabtay) [inline Stresser/ProcessArtifact.h:19]: Use override keyword 
(And install ReSharper 🥇 already ! Hhhh)

File: Utils/LocalPcUtils.cpp [hunk 9]
```diff
@@ -24,7 +24,7 @@ bool LocalPcUtils::doesFileExists(const std::wstring& path)
 
 std::wstring LocalPcUtils::getTempPath()
 {
-	std::vector<WCHAR> tempPath[MAX_PATH + 1];
+	std::vector<WCHAR> tempPath(MAX_PATH + 1);
```
Comment (AvivShabtay) [inline Utils/LocalPcUtils.cpp:27]: 👍

**Review by AvivShabtay — APPROVED** [review]
Very good jobs.
Few reviews on conventions that's all.
Thanks a lot

File: Stresser/ProcessArtifact.cpp (hunk 4 above)
Comment (EladShriki) [reply Stresser/ProcessArtifact.cpp]: DWORD is unsigned long

**Pull request merged**